src/sentry/utils/retries.py

```diff
@@ -0,0 +1,43 @@
+from __future__ import absolute_import
+
+import itertools
+import logging
+import random
+import time
+
+
+logger = logging.getLogger(__name__)
+
+
+class RetryPolicy(object):
+    def __call__(self, function):
+        raise NotImplementedError
+
+
+class TimedRetryPolicy(RetryPolicy):
+    def __init__(self, timeout, delay=None, exceptions=(Exception,)):
+        if delay is None:
+            delay = lambda i: 0.1 + ((random.random() - 0.5) / 10)
+
+        self.timeout = timeout
+        self.delay = delay
+        self.exceptions = exceptions
+
+    def __call__(self, function):
+        start = time.time()
+        for i in itertools.count(1):
+            try:
+                return function()
+            except self.exceptions as error:
+                delay = self.delay(i)
+                if (time.time() + delay) > (start + self.timeout):
+                    raise Exception('Could not successfully execute %r within %.3f seconds (%s attempts.)' % (function, time.time() - start, i))
```

Comment:
Should memoize `time.time()` result so it's exactly the same in the exception vs what it was when we compared it.